third_party/move/move-model/src/builder/exp_builder.rs

```diff
@@ -1681,6 +1760,131 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
                                     RewriteResult::Rewritten(self.new_error_exp().into_exp())
                                 }
                             },
+                            ExpPlaceholder::ReceiverCallInfo {
+                                name,
+                                generics,
+                                arg_types,
+                                result_type,
+                            } => {
+                                // Clone info to avoid borrowing conflicts
+                                let (name, generics, mut arg_types, result_type) = (
+                                    *name,
+                                    generics.clone(),
+                                    arg_types.clone(),
+                                    result_type.clone(),
+                                );
+                                let receiver_arg_ty = self.subs.specialize(
+                                    arg_types
+                                        .first()
+                                        .expect("receiver has at least one argument"),
+                                );
+                                if let Some(inst) =
+                                    self.get_receiver_function(&receiver_arg_ty, name)
+                                {
+                                    let receiver_param_type =
+                                        inst.arg_types.first().expect("argument").clone();
+                                    // Determine whether an automatic borrow needs to be inserted
+                                    // and it's kind.
+                                    let borrow_kind_opt =
+                                        inst.receiver_needs_borrow(&receiver_arg_ty);
+                                    if !inst.type_inst.is_empty() {
+                                        // We need to annotate the instantiation of the function
+                                        // at the node. To obtain it, unification needs to be run
+                                        // again. If unification fails, errors will have been
+                                        // already reported, so we can ignore the result.
+                                        let mut subs = self.subs.clone();
+                                        let mut ok = true;
```

Comment:
This code is pretty unreadable.  It seems like it would be so much better to put it in a helper function (or even in a new method of `Substution` so code can be clarified by early returns and the `?` operator, e.g.:
```
let unify_instantiation = |subs|: Result<Subs, TypeUnificationError> = {
    if let Some(tys) = generics {
        subs.unify_vec_maybe_type_args(...)?;
    }
    if let Some(ref_kind) = &borrow_kind_opt {
        ...
    }
    subs.unify_vec(...)?;
    subs.unify(...)?;
    OK(subs)
}
if let Ok(new_subs) = unify_instantiation(self.subs.clone()) {
    self.subs = new_subs;
    ...
```
or something similar.

Reply:
I pulled this code into its own method and refactored it a bit.